ros2_foxglove_bridge/src/ros2_foxglove_bridge.cpp

```diff
@@ -1,24 +1,343 @@
+    subscriptionOptions.event_callbacks = eventCallbacks;
+
+    constexpr size_t QUEUE_LENGTH = 10;
+    try {
+      auto subscriber = this->create_generic_subscription(
+        topic, datatype, QUEUE_LENGTH,
+        [&](std::shared_ptr<rclcpp::SerializedMessage> msg) {
+          rosMessageHandler(channel, msg);
+        },
+        subscriptionOptions);
+      _subscriptions.emplace(channelId, std::move(subscriber));
+      RCLCPP_INFO(this->get_logger(), "Subscribed to topic \"%s\" (%s)", topic.c_str(),
+                  datatype.c_str());
+    } catch (const std::exception& ex) {
+      RCLCPP_ERROR(this->get_logger(), "Failed to subscribe to topic \"%s\" (%s): %s",
+                   topic.c_str(), datatype.c_str(), ex.what());
+    }
+  }
+
+  void unsubscribeHandler(foxglove::ChannelId channelId) {
+    std::lock_guard<std::mutex> lock(_subscriptionsMutex);
+
+    auto it = _channelToTopicAndDatatype.find(channelId);
+    TopicAndDatatype topicAndDatatype =
+      it != _channelToTopicAndDatatype.end()
+        ? it->second
+        : std::make_pair<std::string, std::string>("[Unknown]", "[Unknown]");
+
+    auto it2 = _subscriptions.find(channelId);
+    if (it2 == _subscriptions.end()) {
+      RCLCPP_WARN(this->get_logger(), "Received unsubscribe request for unknown channel %d",
+                  channelId);
+      return;
+    }
+
+    RCLCPP_INFO(this->get_logger(), "Unsubscribing from topic \"%s\" (%s)",
+                topicAndDatatype.first.c_str(), topicAndDatatype.second.c_str());
+    _subscriptions.erase(it2);
+  }
+
+  void logHandler(LogLevel level, char const* msg) {
+    switch (level) {
+      case LogLevel::Debug:
+        RCLCPP_DEBUG(this->get_logger(), "[WS] %s", msg);
+        break;
+      case LogLevel::Info:
+        RCLCPP_INFO(this->get_logger(), "[WS] %s", msg);
+        break;
+      case LogLevel::Warn:
+        RCLCPP_WARN(this->get_logger(), "[WS] %s", msg);
+        break;
+      case LogLevel::Error:
+        RCLCPP_ERROR(this->get_logger(), "[WS] %s", msg);
+        break;
+      case LogLevel::Critical:
+        RCLCPP_FATAL(this->get_logger(), "[WS] %s", msg);
+        break;
+    }
+  }
+
+  void rosMessageHandler(const foxglove::Channel& channel,
+                         std::shared_ptr<rclcpp::SerializedMessage> msg) {
+    // NOTE: Do not call any RCLCPP_* logging functions from this function. Otherwise, subscribing
+    // to `/rosout` will cause a feedback loop
+    auto timestamp = this->now().nanoseconds();
+    auto payload =
+      std::string_view{reinterpret_cast<const char*>(msg->get_rcl_serialized_message().buffer),
+                       msg->get_rcl_serialized_message().buffer_length};
+    _server.sendMessage(channel.id, timestamp, payload);
+  }
 };
 
 int main(int argc, char* argv[]) {
   rclcpp::init(argc, argv);
-  rclcpp::spin(std::make_shared<FoxgloveBridge>());
+  auto bridgeNode = std::make_shared<FoxgloveBridge>();
+  rclcpp::spin(bridgeNode);
+  bridgeNode.reset();
```

Comment:
this change shouldn't change anything?

Reply:
It explicitly calls the `FoxgloveBridge` destructor to start the process of sending TCP connection close packets to connected clients as early as possible before we go into the ROS 2 shutdown. It might not make any difference, I need to test still.

Reply:
If anything it would only _extend_ the lifetime of the FoxgloveBridge object (depending on whether spin() does anything to remove its strong reference before it returns)